flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkBushyJoinReorderRule.java

```diff
@@ -0,0 +1,620 @@
+ * keep the best plan (have the lowest {@link FlinkCost}) for the same set of m inputs. E.g., for
+ * three-inputs join, we keep only the best plan for inputs {A, B, C} among plans (A J B) J C, (A J
+ * C) J B, (B J C) J A.
+ *
+ * <p>Second, we will add all outer join type inputs in the MultiJoin on the top.
+ */
+public class FlinkBushyJoinReorderRule extends RelRule<FlinkBushyJoinReorderRule.Config>
+        implements TransformationRule {
+
+    /** Creates an SparkJoinReorderRule. */
+    protected FlinkBushyJoinReorderRule(Config config) {
+        super(config);
+    }
+
+    @Deprecated // to be removed before 2.0
+    public FlinkBushyJoinReorderRule(RelBuilderFactory relBuilderFactory) {
+        this(Config.DEFAULT.withRelBuilderFactory(relBuilderFactory).as(Config.class));
+    }
+
+    @Deprecated // to be removed before 2.0
+    public FlinkBushyJoinReorderRule(
+            RelFactories.JoinFactory joinFactory,
+            RelFactories.ProjectFactory projectFactory,
+            RelFactories.FilterFactory filterFactory) {
+        this(RelBuilder.proto(joinFactory, projectFactory, filterFactory));
+    }
+
+    @Override
+    public void onMatch(RelOptRuleCall call) {
+        final RelBuilder relBuilder = call.builder();
+        final MultiJoin multiJoinRel = call.rel(0);
+        final LoptMultiJoin multiJoin = new LoptMultiJoin(multiJoinRel);
+        RelNode bestOrder = findBestOrder(call.getMetadataQuery(), relBuilder, multiJoin);
+        call.transformTo(bestOrder);
+    }
+
+    private static RelNode findBestOrder(
+            RelMetadataQuery mq, RelBuilder relBuilder, LoptMultiJoin multiJoin) {
+        // In our bushy join reorder strategy, we will first try to reorder all the inner join type
+        // inputs in the multiJoin, and then add all outer join type inputs on the top.
+        // First, reorder all the inner join type inputs in the multiJoin.
+        List<Map<Set<Integer>, JoinPlan>> foundPlans = reOrderInnerJoin(mq, relBuilder, multiJoin);
+
+        JoinPlan finalPlan;
+        // Second, add all outer join type inputs in the multiJoin on the top.
+        if (canOuterJoin(multiJoin)) {
+            finalPlan =
+                    addToTopForOuterJoin(
+                            getBestPlan(mq, foundPlans.get(foundPlans.size() - 1)),
+                            multiJoin,
+                            relBuilder);
+        } else {
+            if (foundPlans.size() != multiJoin.getNumJoinFactors()) {
+                finalPlan =
+                        addToTop(
+                                getBestPlan(mq, foundPlans.get(foundPlans.size() - 1)),
+                                multiJoin,
+                                relBuilder);
+            } else {
+                assert foundPlans.get(foundPlans.size() - 1).size() == 1;
+                finalPlan = new ArrayList<>(foundPlans.get(foundPlans.size() - 1).values()).get(0);
+            }
+        }
```

Comment:
can you explain this part, it's a little confuse

Reply:
> can you explain this part, it's a little confuse

Done!